### ISSUE TITLE
Use `Atomics.waitAsync` rather than callbacks once stable

### ISSUE THREAD

**Issue by tomaka** [smol-dot/smoldot#1128]
The only browser not supporting `waitAsync` is Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1467846

Once Firefox ships this function, we can replace the JSON-RPC response notification and executor ready notification callbacks to notify a futex instead.

This would both improve the code readability and be more optimal.

**Comment by tomaka** [comment]
I've just realized that this requires the `atomics` target feature, which is annoying from the Rust side because enabling this feature hackily enables thread safety in code.

**Comment by tomaka** [comment]
> I've just realized that this requires the atomics target feature, which is annoying from the Rust side because enabling this feature hackily enables thread safety in code.

I'm going to close this issue for this reason.

As explained in #91, I don't think that we want multithreading-friendly wasm.
As such, we would need to enable the `atomics` feature just to be able to do what this issue is about.
Enabling this feature would reduce the platforms smoldot can run on, but more important is pretty annoying to do in Rust right now due to the standard library treating `atomics` as a special feature that requires recompiling it.

In other words, since it's just to make the code nicer, it's not worth the added complexity. But it could become worth in the future if Rust ships a stdlib with `atomics` enabled.
